### ISSUE TITLE
Issue reading tags of the predefined type Message

### ISSUE THREAD

**Issue by jggri** [libplctag/libplctag#505]
Hi,

I am trying to read this Controller tag
![imagen](https://github.com/user-attachments/assets/3da2740a-4246-4910-b2d2-384122e71f41)
The type MESSAGE is a predefined type with the following definition
![imagen](https://github.com/user-attachments/assets/75394a5e-1e08-451d-adc9-36187e33af57)

If I use the tag name (in this case, ReadEthernetMSG), it doesn’t work. I get the error PLCTAG_ERR_NOT_ALLOWED, and when I analyze it with Wireshark, I see General status -> Privilege violation. However, if I access the tag using the field name, for example, ReadEthernetMSG.Flags, it works.

Is there a special format or configuration required to read this type? I am interested in reading all the field simultaneously

Thanks.

**Comment by jggri** [comment]
2024-12-09 13:24:00.420 thread(1) tag(0) WARN receive_forward_open_response:2567 Forward Open command failed, response code: Unsupported service. (8)
2024-12-09 13:24:00.420 thread(1) tag(0) WARN receive_forward_open_response:2570 Received CIP command unsupported error from the PLC!
2024-12-09 13:24:00.829 thread(2) tag(0) WARN check_read_status_connected:1394 CIP read failed with status: 0xf Permission denied.
2024-12-09 13:24:00.829 thread(2) tag(0) WARN check_read_status_connected:1525 Error received!
2024-12-09 13:24:00.829 thread(3) tag(11) WARN plc_tag_create_ex:1056 Error PLCTAG_ERR_NOT_ALLOWED while trying to create tag!
2024-12-09 13:24:00.991 thread(3) tag(11) WARN recv_eip_response:2298 Timed out waiting for data to read!
2024-12-09 13:24:00.991 thread(3) tag(11) WARN recv_forward_close_resp:2689 Unable to receive Forward Close response, PLCTAG_ERR_TIMEOUT!
2024-12-09 13:24:00.991 thread(3) tag(11) WARN perform_forward_close:2338 Forward close response not received, PLCTAG_ERR_TIMEOUT!

**Comment by jggri** [comment]
External Access = Read/Write

**Comment by kyle-github** [comment]
Thanks for using libplctag.   Can you please set debug level to 4 and capture the log so that I can see the bytes that are coming back from the PLC?   Something looks a bit fishy here.  The code might be translating the error incorrectly.   Normally an error of 0x08 (unsupported command) results in trying the old ForwardOpen, not in a privilege violation.

**Comment by jggri** [comment]
Thanks kyle. Attached logs and wireshark 
![imagen](https://github.com/user-attachments/assets/f9657d00-6407-4d51-a0ad-1c5c5eb2942a)
![imagen](https://github.com/user-attachments/assets/931aee50-c0c3-44ec-b3f4-b8ee243675fd)
[log_reading_field_ok.txt](https://github.com/user-attachments/files/18061500/log_reading_field_ok.txt)
[log_reading_tag.txt](https://github.com/user-attachments/files/18061501/log_reading_tag.txt)

**Comment by kyle-github** [comment]
Thanks for the logs.   The PLC is definitely failing the CIP command in the second log file.

According to my translation table error code 0x0f means:

```
0x0F - Privilege violation - A permission/privilege check failed
```

The CIP response is this:

```
d2 00 0f 00
```

0xd2 is the byte that says that the read is done (0x80 | 0x52).   0x0f is the status byte.

So the PLC really is responding with an error that there is a privilege violation.   That is strange that the privs seem to be set for Read/Write and it is failing.   Is there some other layer of protection?

I see nothing wrong with the tag read sent to the PLC.  It all looks good.

**Comment by jggri** [comment]
Yes, is very strange, because some predefined types work and others don't (same configuration and privileges).  I have checked with RSLinx Classic Lite -> Data Monitor, and it works. It might be using a different access method, or it could be accessing each field individually for every read (this approach works with the library using tagName.tagFieldName
![imagen](https://github.com/user-attachments/assets/23a735b3-bb07-4255-97e5-8b3784dbeb67)

**Comment by jggri** [comment]
I have tested all the predefined types available in Studio 5000, and I encounter the same issue with the following types:

    ALARM_ANALOG
    ALARM_DIGITAL
    AXIS_CIP_DRIVE
    COORDINATE_SYSTEM
    MESSAGE
    MOTION_GROUP

getting a privilege violation error when using the tag name directly

**Comment by jggri** [comment]
The library is working fine

**Comment by kyle-github** [comment]
I found a PLC with a program that has MESSAGE tags and I get exactly the same error.   It isn't your program.   I've dug into the response from the PLC and it is definitely a privilege error that is being returned.

**Comment by kyle-github** [comment]
This is really a concern!   If you don't mind I would like to reopen this as a bug/enhancement to track what is learned.

**Comment by jggri** [comment]
A solution to this issue would be to read field by field for those tags where reading them by their name results in a Privilege Violation. I've checked, it works for the previouslyt predefined types

**Comment by kyle-github** [comment]
That is a possibility but I currently have the library set up to _not_ list out tag definitions automatically and it has no way to use them internally.   That said, I strongly suspect that the Rockwell tools must use some other way to read this data.

**Comment by jggri** [comment]
Yes, it uses a different way to read the data when using RSLinx Classic Lite -> Data Monitor

In the document https://literature.rockwellautomation.com/idc/groups/literature/documents/pm/1756-pm020_-en-p.pdf 
![imagen](https://github.com/user-attachments/assets/da35750c-0f54-4014-a7e5-1bc4d88e202a)

It states that Symbol Instance Addressing uses class 0x6B for both Read Tag Service and Read Tag Fragmented Service
![imagen](https://github.com/user-attachments/assets/b68e1538-bc58-491f-9974-be6a741c78aa)

However, after doing reverse ingeneering to determine the way it uses to read the tag value it is not using the methods explained in the document. I created a DINT tag with a value 1431655765 = 0x55 55 55 55. I can see it is sent using this

![imagen](https://github.com/user-attachments/assets/8cb00f78-6df4-45a2-8d99-2e0f3a96cd97)

And the library uses 
![imagen](https://github.com/user-attachments/assets/62e766f5-55c9-497a-a1ca-c2e3f89997c5)

**Comment by kyle-github** [comment]
The library uses the symbolic/name-based method to access tags.  There are ways to access tags based on the class and instance IDs of the tags.  Those are not supported by the library yet.  Sounds like they need to be!

**Comment by kyle-github** [comment]
This depends on #200.

I am going to close this one for now as #200 already has most of the underlying requirements to fix this.
